### PR TITLE
Add a new tool for testing provisioner installations like a real user

### PR DESCRIPTION
**Pull request by spirosoik**
#### Summary
The tool is called testwick and can be build by itself as a CLI tool. The
implementation though is like a framework which you can use by itself and
write your own scenarios but of course always respecting the order of the
actions.

Right now it creates installations and tear down them in the end. It runs
couple of actions to these installations like:
- Create Channel
- Create Users
- Create Teams
- Add team members
- Post messages

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-37408

#### Release Note
```release-note
Adding a new isolated tool called `testwick` and it is used to create multiple installations and interact with them
```

### CODE REVIEW

**Comment by spirosoik** [comment]
@Szymongib it's almost done but I need to add the test for testwicker. Though before open the PR feedback will be very valuable.

File: cmd/tools/testwick/go.mod [hunk 1]
```diff
@@ -0,0 +1,13 @@
+module github.com/mattermost/mattermost-cloud/cmd/tools/testwick
```
Comment (Szymongib) [inline cmd/tools/testwick/go.mod:1]: I wonder what are advantages of adding it as a separate module?

File: cmd/tools/testwick/testwicker.go [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+// PostMessage post messages to the different channels which are created
+// automatically by the wicker
+func (w *TestWicker) PostMessage(samples int) *TestWicker {
+	w.logger.WithField("DNS", w.installation.DNS).Info("Posting messages")
+	if w.channelID == "" {
+		w.Error = fmt.Errorf("failed to post message. You need to create channels first")
+		return w
+	}
+	for i := 0; i < samples; i++ {
+		_, response := w.mmClient.CreatePost(&mmodel.Post{
```
Comment (Szymongib) [inline cmd/tools/testwick/testwicker.go:77]: Did not know there is a MM client like that 😄  +1 for using it.

File: cmd/tools/testwick/testwicker.go [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+	if response.StatusCode != 201 {
+		w.Error = fmt.Errorf("failed to create channel status = %d, message = %s", response.StatusCode, response.Error.Message)
+		w.logger.WithError(w.Error).Error()
+		return w
+	}
+	return w
+}
+
+// WaitForInstallationStable waits for installation to become stable
+func (w *TestWicker) WaitForInstallationStable(ctx context.Context) *TestWicker {
```
Comment (Szymongib) [inline cmd/tools/testwick/testwicker.go:170]: We have some similar functions in our e2e tests https://github.com/mattermost/mattermost-cloud/blob/master/e2e/pkg/installation.go#L52
I think it might be worth it to standardize it down the line in some package, but that is outside of the scope of this PR.

**Review by Szymongib — COMMENTED** [review]
Looks really good at a first glance! Great work!
 Ping me when ready to final review 😉

File: cmd/tools/testwick/go.mod (hunk 1 above)
Comment (spirosoik) [reply cmd/tools/testwick/go.mod:1]: avoid bumping aws sdk version etc in the main go module

File: cmd/tools/testwick/testwicker.go (hunk 2 above)
Comment (spirosoik) [reply cmd/tools/testwick/testwicker.go:77]: yeah which is amazing actually :)

**Comment by spirosoik** [comment]
@Szymongib I think to write the unit will be an overkill as I need to create a mock for the whole mattermost api client as it doesn't exist as I see.

File: cmd/tools/testwick/testwicker.go [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+// TestWicker data struct for test wicker scenarios
+type TestWicker struct {
+	provisionerClient *cmodel.Client
+	mmClient          *mmodel.Client4
+	logger            *logrus.Logger
+	installation      *cmodel.InstallationDTO
+	userID            string
+	channelID         string
+	teamID            string
+	Error             error
```
Comment (Szymongib) [inline cmd/tools/testwick/testwicker.go:28]: I am not sure about error handling here. Let's consider the following scenario:
- `CreateInstallation` failes
- `Error` is set to some value but the flow continues
- `WaitForInstallationStable` will also fail as the installation was not created
- It will override `Error` and the process will continue most likely failing further along the way

If I am not missing anything this will cause only the last error to be logged therefore it may be hard to troubleshoot.

File: cmd/tools/testwick/testwicker.go [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+		w.Error = fmt.Errorf("failed logging user: username = %s, status code = %d, message = %s", u.Username, response.StatusCode, response.Error.Message)
+		return w
+	}
+	w.userID = userLogged.Id
+	return w
+}
+
+// IsUP cpings the installation till return statusCode=200 and status=OK
+func (w *TestWicker) IsUP(ctx context.Context) *TestWicker {
+	w.mmClient = mmodel.NewAPIv4Client(fmt.Sprintf("https://%s", w.installation.DNS))
```
Comment (Szymongib) [inline cmd/tools/testwick/testwicker.go:227]: Could we initialize the client earlier? I think we already have a DNS prior to the creation of installation, and initializing it here might lead to some confusion/errors if we will be extending the code, adding new methods, and mixing things up.

**Review by Szymongib — COMMENTED** [review]
@spirosoik I think you could just define an interface here with methods you need and mock only those methods. That is pretty common practice in Go.

That said I am fine with not including unit tests for now as this is test code actually and there is not much logic 😄  If we will have time down the line it may be beneficial but should be good for now.

File: cmd/tools/testwick/testwicker.go (hunk 5 above)
Comment (spirosoik) [reply cmd/tools/testwick/testwicker.go:227]: sure agree

File: cmd/tools/testwick/testwicker.go (hunk 4 above)
Comment (spirosoik) [reply cmd/tools/testwick/testwicker.go:28]: Yeah this is true, I think I can split it to `testwicker, error`

**Comment by spirosoik** [comment]
> @spirosoik I think you could just define an interface here with methods you need and mock only those methods. That is pretty common practice in Go.
> 
> That said I am fine with not including unit tests for now as this is test code actually and there is not much logic 😄 If we will have time down the line it may be beneficial but should be good for now.

yeah I totally forgot the interface as I struggle with nginx in community :P !!! I will add the tests it's pretty easy actually and I will pass the interface in the `New...` factory method and all good

File: cmd/tools/testwick/testwicker.go (hunk 4 above)
Comment (spirosoik) [reply cmd/tools/testwick/testwicker.go:28]: @Szymongib what if we just keep a list of errors and dump them as logs if `len(w.Error) > 0`. Trying to think how we can keep the declarative usage but still keep all the logs.

Comment (spirosoik) [reply cmd/tools/testwick/testwicker.go:28]: Or even better have a `Runner()` which gets as input functions and iterate through them:

```go
runner.Execute([func1, func2])
```

internally the runner checks for errors and fails accordingly so we can avoid the multiple `if err!=nil` for every step

Comment (Szymongib) [reply cmd/tools/testwick/testwicker.go:28]: That sounds good, I think we actually may have something similar in e2e tests https://github.com/mattermost/mattermost-cloud/blob/master/e2e/workflow/engine.go#L30

**Comment by spirosoik** [comment]
@Szymongib when you have time check the workflow runner, I created my own as it's slight different and I don't need the in parallel queue for now etc.

**Review by Szymongib — APPROVED** [review]
LGTM! Great stuff!
